client/dashboard/sites/site/environment-switcher.tsx

```diff
@@ -0,0 +1,167 @@
+import { useQuery, useMutation } from '@tanstack/react-query';
+import {
+	__experimentalHStack as HStack,
+	Button,
+	Dropdown,
+	MenuGroup,
+	MenuItem,
+	Spinner,
+} from '@wordpress/components';
+import { useDispatch } from '@wordpress/data';
+import { sprintf, __ } from '@wordpress/i18n';
+import { Icon, chevronDownSmall, plus } from '@wordpress/icons';
+import { store as noticesStore } from '@wordpress/notices';
+import { siteByIdQuery } from '../../app/queries/site';
+import { stagingSiteCreateMutation } from '../../app/queries/site-staging-sites';
+import { production, staging } from '../../components/icons';
+import RouterLinkMenuItem from '../../components/router-link-menu-item';
+import { canManageSite, canCreateStagingSite } from '../features';
+import type { Site } from '../../data/types';
+
+type EnvironmentType = 'production' | 'staging';
+
+const Environment = ( { env }: { env: EnvironmentType } ) => {
+	if ( env === 'staging' ) {
+		return (
+			<HStack justify="flex-start" style={ { width: 'auto', flexShrink: 0 } }>
+				<Icon icon={ staging } />
+				<span>{ __( 'Staging' ) }</span>
+			</HStack>
+		);
+	}
+
+	return (
+		<HStack justify="flex-start" style={ { width: 'auto', flexShrink: 0 } }>
+			<Icon icon={ production } />
+			<span>{ __( 'Production' ) }</span>
+		</HStack>
+	);
+};
+
+const CurrentEnvironment = ( { site }: { site: Site } ) => {
+	if ( site.is_wpcom_staging_site ) {
+		return <Environment env="staging" />;
+	}
+
+	return <Environment env="production" />;
+};
+
+const EnvironmentSwitcherDropdown = ( {
+	currentSite,
+	otherEnvironment,
+	otherEnvironmentSite,
+	onClose,
+}: {
+	currentSite: Site;
+	otherEnvironment: EnvironmentType;
+	otherEnvironmentSite?: Site;
+	onClose: () => void;
+} ) => {
+	const productionSite = otherEnvironment === 'staging' ? currentSite : otherEnvironmentSite;
+	const stagingSite = otherEnvironment === 'staging' ? otherEnvironmentSite : currentSite;
+	const { createSuccessNotice, createErrorNotice } = useDispatch( noticesStore );
+	const mutation = useMutation( stagingSiteCreateMutation( productionSite?.ID ?? 0 ) );
+	const handleCreate = () => {
+		mutation.mutate( undefined, {
+			onSuccess: () => {
+				createSuccessNotice( __( 'Staging site created.' ), { type: 'snackbar' } );
+			},
+			onError: ( error: Error ) => {
+				createErrorNotice(
+					sprintf(
+						// translators: "reason" is why adding the staging site failed.
+						__( 'Failed to create staging site: %(reason)s' ),
+						{ reason: error.message }
+					),
+					{
+						type: 'snackbar',
+					}
+				);
+			},
+		} );
+	};
+
+	// TODO: Handle upsell.
+	const handleUpsell = () => {};
+
+	return (
+		<MenuGroup>
+			{ productionSite && canManageSite( productionSite ) && (
```

Comment:
Should we remove all `canManageSite()` checks in the PR? Because we already check at the `siteRoute` level 🤔 

https://github.com/Automattic/wp-calypso/blob/61336d415a49596c031c3958c538b3170bf8f5d3/client/dashboard/sites/site/index.tsx#L21-L23

Reply:
Oops... not always. We can invite another user to staging sites but not to the production sites 😄 🤦